Include typed dependencies in `mypy` pre-commit checks

@grovduck discovered in #114 that the current `mypy` checks weren't catching some clear typing errors related to Numpy and Pandas types, which stems from the fact that our pre-commit environment is too isolated.

From the [mypy pre-commit readme](https://github.com/pre-commit/mirrors-mypy/tree/main?tab=readme-ov-file#mypy-mirror):

> Because pre-commit runs mypy from an isolated virtualenv (without your dependencies) you may also find it useful to add the typed dependencies to additional_dependencies so mypy can better perform dynamic analysis

Using the pre-commit config below provides more thorough checks, which uncovers a number of existing typing errors that weren't previously being caught (~9 errors on `main`~ 5 errors on `main` after merging #109 and #110).

```yaml
  - repo: https://github.com/pre-commit/mirrors-mypy
    rev: v1.19.1
    hooks:
      - id: mypy
        exclude: ^tests/
        additional_dependencies:
          - numpy
          - pandas-stubs
          - scipy-stubs
```

The tricky question is what order to address the different typing changes in, given that the existing PRs #109, #110, and #112 may resolve some of the underlying typing issues, but may also add some of their own, since they weren't being thoroughly checked.

Let me know what you think @grovduck, but I'm tempted to continue going through those PRs in order, temporarily using the expanded configuration above to make sure they don't add any new typing errors, but ignoring any existing errors. Once those are merged in, we can permanently update the configuration and tackle the pre-existing typing errors that are still left. I'm hoping that would minimize the amount of rebasing needed, without creating a bunch of new errors that we need to revert and fix.

<details>
<summary>Type errors on `main`</summary>

```
src/sknnr/datasets/_base.py:219: error: Incompatible types in assignment (expression has type "Dataset[DataFrame]", variable has type "Dataset[ndarray[tuple[Any, ...], dtype[float64]]]")  [assignment]
src/sknnr/transformers/_tree_node_transformer.py:72: error: Argument 1 to "issubdtype" has incompatible type "dtype[Any] | CategoricalDtype"; expected "type[Any] | dtype[Any] | _HasDType[dtype[Any]] | _HasNumPyDType[dtype[Any]] | tuple[Any, Any] | list[Any] | _DTypeDict | str | None"  [arg-type]
src/sknnr/_weighted_trees.py:39: error: Bad number of arguments for type alias, expected 0, given 1  [type-arg]
src/sknnr/_rfnn.py:181: error: Bad number of arguments for type alias, expected 0, given 1  [type-arg]
src/sknnr/_gbnn.py:183: error: Bad number of arguments for type alias, expected 0, given 1  [type-arg]
Found 5 errors in 5 files (checked 25 source files)
```

</details>

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -6,7 +6,7 @@ repos:
       - id: ruff-format
 
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v1.19.1
+    rev: v1.20.2
     hooks:
       - id: mypy
         exclude: ^tests/

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -10,4 +10,7 @@ repos:
     hooks:
       - id: mypy
         exclude: ^tests/
-        args: [--ignore-missing-imports]
+        additional_dependencies:
+          - numpy
+          - pandas-stubs
+          - scipy-stubs

--- a/src/sknnr/_gbnn.py
+++ b/src/sknnr/_gbnn.py
@@ -182,7 +182,7 @@ class GBNNRegressor(WeightedTreesNNRegressor):
         n_iter_no_change: int | None = None,
         tol: float = 0.0001,
         ccp_alpha: float = 0.0,
-        forest_weights: Literal["uniform"] | ArrayLike[float] = "uniform",
+        forest_weights: Literal["uniform"] | ArrayLike = "uniform",
         tree_weighting_method: Literal[
             "train_improvement", "uniform"
         ] = "train_improvement",

--- a/src/sknnr/_rfnn.py
+++ b/src/sknnr/_rfnn.py
@@ -180,7 +180,7 @@ class RFNNRegressor(WeightedTreesNNRegressor):
         ccp_alpha: float = 0.0,
         max_samples: int | float | None = None,
         monotonic_cst: list[int] | None = None,
-        forest_weights: Literal["uniform"] | ArrayLike[float] = "uniform",
+        forest_weights: Literal["uniform"] | ArrayLike = "uniform",
         n_neighbors: int = 5,
         weights: Literal["uniform", "distance"] | Callable = "uniform",
     ):

--- a/src/sknnr/_weighted_trees.py
+++ b/src/sknnr/_weighted_trees.py
@@ -41,7 +41,7 @@ class WeightedTreesNNRegressor(YFitMixin, TransformedKNeighborsRegressor):
     """
 
     transformer_: TreeNodeTransformer
-    forest_weights: Literal["uniform"] | ArrayLike[float]
+    forest_weights: Literal["uniform"] | ArrayLike
 
     def __init__(
         self,

--- a/src/sknnr/datasets/_base.py
+++ b/src/sknnr/datasets/_base.py
@@ -207,7 +207,7 @@ def load_dataset_from_csv_filenames(
         file_name=target_filename, module_name=module_name
     )
 
-    dataset = Dataset(
+    array_ds = Dataset(
         index=index,
         data=data,
         target=target,
@@ -217,9 +217,10 @@ def load_dataset_from_csv_filenames(
     )
 
     if as_frame:
-        dataset = _dataset_as_frame(dataset)
+        frame_ds = _dataset_as_frame(array_ds)
+        return (frame_ds.data, frame_ds.target) if return_X_y else frame_ds
 
-    return (dataset.data, dataset.target) if return_X_y else dataset
+    return (array_ds.data, array_ds.target) if return_X_y else array_ds
 
 
 @overload

